### ISSUE TITLE
Training Stops with Loss=Nan at a certain epoch. 

### ISSUE THREAD

**Issue by pseulki** [WangWenhao0716/TransHP#5]
Thank you for sharing such great work! 

I'm running it on the iNaturalist dataset, but I've encountered an issue where the training consistently stops at a certain epoch (e.g., 109th epoch) because the sig_loss[x] for the coarse prompt-related loss turns into NaN. I tried lowering the learning rate, but strangely, it still stops at that specific epoch.

Looking at engine.py, I noticed that there's explicit code to check if the loss_value is finite, which makes me wonder if you've experienced this issue during training. Do you have any recommendations for getting past the NaN issue? 

Thanks!

**Comment by WangWenhao0716** [comment]
Hi, it is a known issue caused by DeiT. See: https://github.com/facebookresearch/deit/issues/29

**Comment by pseulki** [comment]
Thanks for your response! :) 
I saw in the thread that using a smaller learning rate is one recommendation. When using your method as a baseline, would it be okay to try smaller learning rates, like 1e-4 or 5e-5, instead of the 5e-4 you used, to avoid NaNs after that specific epoch?

**Comment by WangWenhao0716** [comment]
No, do not change the learning rate. Change the attention layer from FP16 to FP32z

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by WangWenhao0716** [comment]
> Thanks for your response! :) I saw in the thread that using a smaller learning rate is one recommendation. When using your method as a baseline, would it be okay to try smaller learning rates, like 1e-4 or 5e-5, instead of the 5e-4 you used, to avoid NaNs after that specific epoch?

You should try:

```python
class Attention(nn.Module):
    def __init__(self, dim, num_heads=8, qkv_bias=False, qk_scale=None, attn_drop=0., proj_drop=0.):
        super().__init__()
        self.num_heads = num_heads
        head_dim = dim // num_heads
        # NOTE scale factor was wrong in my original version, can set manually to be compat with prev weights
        self.scale = qk_scale or head_dim ** -0.5

        self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
        self.attn_drop = nn.Dropout(attn_drop)
        self.proj = nn.Linear(dim, dim)
        self.proj_drop = nn.Dropout(proj_drop)

    def forward(self, x):
        B, N, C = x.shape
        qkv = self.qkv(x).reshape(B, N, 3, self.num_heads, C // self.num_heads).permute(2, 0, 3, 1, 4)

        with torch.cuda.amp.autocast(enabled=False):
            q, k, v = qkv[0].float(), qkv[1].float(), qkv[2].float()   # make torchscript happy (cannot use tensor as tuple)
            attn = (q @ k.transpose(-2, -1)) * self.scale
            attn = attn.softmax(dim=-1)
            attn = self.attn_drop(attn)
            x = (attn @ v).transpose(1, 2).reshape(B, N, C)

        x = self.proj(x)
        x = self.proj_drop(x)
        return x
```

**Comment by pseulki** [comment]
I'll try that. Thank you for the detailed guide :)

**Comment by pseulki** [comment]
Hi, I tried the code (changing attention layer to float), but it doesn't work :'( .

**Comment by WangWenhao0716** [comment]
Oh, I am sorry to hear that. But, I do not have time to solve this currently. Could you make sure using the exactly the same environment (torch version and so on) and change the attention layer to FP32 in timm?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by pseulki** [comment]
Yes, I matched the environment, including the Torch version, and switched to FP32. The difference is that I used iNaturalist instead of ImageNet. The iNat dataset uses the same training settings, right?

Anyway, thanks for taking the time to respond!

**Comment by WangWenhao0716** [comment]
yeah. only the balance parameters are different. It does not influence I think

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
